### ISSUE TITLE
examples/generated-source does not work. Regression in preGenerateCommands.

### ISSUE THREAD

**Issue by veelo** [dlang/dub#1474]
### System information

- **dub version**: dub 1.9.0
- **OS Platform and distribution**: Windows 10, and Travis-CI (Linux)
- **compiler version** dmd-2.080.0

### Bug Description
`examples/generated-sources` does not work.

### How to reproduce?
```
git clone https://github.com/dlang/dub.git
cd dub\examples\generated-sources
dub build
```
Produces a link error. A second invocation of `dub` succeeds.

### Expected Behavior
Success upon initial invocation.

### Logs
```
.dub\build\application-debug-windows-x86-dmd_2080-33BFC7CE2FACCDF4772619001CE7CA74\generated-sources.obj(generated-sources)
 Error 42: Symbol Undefined __D4test3funFZi
Error: linker exited with status 1
```

Maybe add this to the `test` directory? (Make sure the generated file is deleted before the test is run.) This is a regression, see issue #144.

**Comment by veelo** [comment]
For reference: forum post and failing use case https://forum.dlang.org/post/zhndrynwhtoryxiykxvy@forum.dlang.org

**Comment by veelo** [comment]
If all sources are generated, the error is different:
`Configuration 'application' of package [...] contains no source files. Please add {"targetType": "none"} to its package description to avoid building it.
Main package must have a binary target type, not none. Cannot build.`
When this issue is fixed, the above case should probably also be tested.

**Comment by wilzbach** [comment]
A PoC fix for this https://github.com/dlang/dub/pull/1708

**Comment by andre2007** [comment]
@wilzbach my assumption is, the use case is wrong. The pre generate commands are executed "always". Generating source code using pre generate commands causing therefore a painful behavior as in the referenced issue from @MikeWey.
The correct use case is in my opinion to use the pre build commands for generating source code. Therefore the build settings needs also be refreshed here too.

**Comment by veelo** [comment]
> @wilzbach my assumption is, the use case is wrong. The pre generate commands are executed "always". Generating source code using pre generate commands causing therefore a painful behavior as in the referenced issue from @MikeWey.
> The correct use case is in my opinion to use the pre build commands for generating source code. Therefore the build settings needs also be refreshed here too.

I don't understand. As I read it, https://github.com/gtkd-developers/GlibD/issues/6#issuecomment-547586145, @MikeWey says the opposite, that using preBuildCommands _won't_ work in a fresh checkout. That's also my experience. You need commands that are _always_ executed, so preGenerateCommands. The commands themselves should be smart enough to not generate code unnecessarily, e.g. `make -f generate.mk`, or `rdmd generate.d` where `generate.d` considers file modification times. This will prevent the painfulness mentioned in https://github.com/gtkd-developers/GlibD/issues/6.

**Comment by andre2007** [comment]
@veelo my understanding is, pregenerate and prebuilt commands both suffers from a similar bug. I would say @MikeWey describes this in the first section. The list of known source files must be refreshed after both execution.

Of course preGenerateCommands can implement a logic to generate code only when necessary. In my opinion this additional logic is superfluous if you use preBuildCommands.

**Comment by s-ludwig** [comment]
@veelo is right here, with the main logical problem being that pre-build commands are potentially out of reach for dub (e.g. when a Visual Studio solution is generated). The only workaround would be to inject something like a `dub generate` command into the solution/make file as a custom build command, which would partially defeat the purpose of `dub generate` in the first place.

So in short, we could make pre-buid commands work for `BuildGenerator`, but for the other generator types it's not really possible in a clean way. As far as I can see, fixing and advocating pre-generate commands is the way to go here.

**Comment by veelo** [comment]
> Of course preGenerateCommands can implement a logic to generate code only when necessary.

How? Dub will need to know on what condition to execute the commands. A proposal: #1857.

**Comment by Geod24** [comment]
Fixed in #2035
